egs/librispeech/asr/simple_v1/ctc_att_transformer_decode.py

```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python3
+
+# Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey, Haowen Qiu)
+#                2021  University of Chinese Academy of Sciences (author: Han Zhu)
+# Apache 2.0
+
+import logging
+import os
+from pathlib import Path
+from typing import List, Optional, Union
+import argparse
+
+import k2
+import torch
+from k2 import Fsa, SymbolTable
+from kaldialign import edit_distance
+from lhotse import CutSet
+from lhotse.dataset.speech_recognition import K2SpeechRecognitionIterableDataset
+
+from snowfall.common import get_phone_symbols
+from snowfall.common import load_checkpoint
+from snowfall.common import average_checkpoint
+from snowfall.common import setup_logger
+from snowfall.common import get_texts
+from snowfall.common import find_first_disambig_symbol
+from snowfall.decoding.graph import compile_LG
+from snowfall.models import AcousticModel
+from snowfall.models.transformer import Transformer
+from snowfall.training.ctc_graph import build_ctc_topo
+
+
+def decode(dataloader: torch.utils.data.DataLoader, model: AcousticModel,
+           device: Union[str, torch.device], LG: Fsa, symbols: SymbolTable):
+    tot_num_cuts = len(dataloader.dataset.cuts)
+    num_cuts = 0
+    results = []  # a list of pair (ref_words, hyp_words)
+    for batch_idx, batch in enumerate(dataloader):
+        feature = batch['features']
+        supervisions = batch['supervisions']
+        supervision_segments = torch.stack(
+            (supervisions['sequence_idx'],
+             torch.floor_divide(supervisions['start_frame'],
+                                model.subsampling_factor),
+             (((supervisions['num_frames'] - 1) // 2 - 1) // 2)), 1).to(torch.int32)
+        indices = torch.argsort(supervision_segments[:, 2], descending=True)
+        supervision_segments = supervision_segments[indices]
+        texts = supervisions['text']
+        assert feature.ndim == 3
+
+        feature = feature.to(device)
+        # at entry, feature is [N, T, C]
+        feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+        with torch.no_grad():
+            nnet_output, _, _ = model(feature, supervision_segments)
+        # nnet_output is [N, C, T]
+        nnet_output = nnet_output.permute(0, 2,
+                                          1)  # now nnet_output is [N, T, C]
+
+        blank_bias = -3.0
```

Comment:
don't know if this was tuned?

Reply:
.. -3.0 might not be optimal for this type of model..

Reply:
I didn't tuned this. Will do some experiments to find optimal bias.